File: java_apx/src/test/java/net/sympower/citizen/apx/DataResponseServiceTest.java
package net.sympower.citizen.apx;

import net.sympower.citizen.apx.config.ProviderEndpointsProperty;
import net.sympower.citizen.apx.dto.outgoing.DataResponse;
import net.sympower.citizen.apx.service.DataRequestService;
import net.sympower.citizen.apx.service.DataResponseService;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

import java.math.BigDecimal;
import java.util.List;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

class DataResponseServiceTest {

    public DataRequestService mockDataRequestService = mock(DataRequestService.class);
    public ProviderEndpointsProperty mockProviderEndpoints = mock(ProviderEndpointsProperty.class);
    public DataResponseService dataResponseService = new DataResponseService(mockDataRequestService, mockProviderEndpoints);

    @BeforeEach
    void setUp() {
        when(mockProviderEndpoints.getSomeProviderGet()).thenReturn("SOME_MOCK_URL");
    }

    @Test
    void shouldGetDataResponseForTheSameQuotes() {
        // GIVEN
        when(mockDataRequestService.getProviderData(any())).thenReturn(MockData.getSameQuotes());
        // WHEN
        List<DataResponse> response = dataResponseService.getDataResponse();
        // THEN
        assertThat(response).hasSize(1);
        assertThat(response.get(0).getMarket()).isEqualTo("SomeMarket");
        assertThat(response.get(0).getDate()).isEqualTo(1573599600000L);
        assertThat(response.get(0).getValues()).hasSize(2);
        assertThat(response.get(0).getValues().get(0).getHour()).isEqualTo(1);
        assertThat(response.get(0).getValues().get(0).getNetVolume()).isEqualTo(BigDecimal.valueOf(11.11));
        assertThat(response.get(0).getValues().get(0).getPrice()).isEqualTo(BigDecimal.valueOf(11.22));
        assertThat(response.get(0).getValues().get(1).getHour()).isEqualTo(2);
        assertThat(response.get(0).getValues().get(1).getNetVolume()).isEqualTo(BigDecimal.valueOf(22.11));
        assertThat(response.get(0).getValues().get(1).getPrice()).isEqualTo(BigDecimal.valueOf(22.22));
    }

    @Test
    void shouldGetDataResponseForDifferentQuotes() {
        // GIVEN
        when(mockDataRequestService.getProviderData(any())).thenReturn(MockData.getDifferentQuotes());
        // WHEN
        List<DataResponse> response = dataResponseService.getDataResponse();
        // THEN
        assertThat(response).hasSize(2);
        assertThat(response.get(0).getMarket()).isEqualTo("SomeMarket");
        assertThat(response.get(0).getDate()).isEqualTo(1573599600000L);
        assertThat(response.get(0).getValues()).hasSize(1);
        assertThat(response.get(0).getValues().get(0).getHour()).isEqualTo(1);
        assertThat(response.get(0).getValues().get(0).getNetVolume()).isEqualTo(BigDecimal.valueOf(11.11));
        assertThat(response.get(0).getValues().get(0).getPrice()).isEqualTo(BigDecimal.valueOf(11.22));
        assertThat(response.get(1).getMarket()).isEqualTo("OtherMarket");
        assertThat(response.get(1).getDate()).isEqualTo(1573599600000L);
        assertThat(response.get(1).getValues()).hasSize(1);
        assertThat(response.get(1).getValues().get(0).getHour()).isEqualTo(3);
        assertThat(response.get(1).getValues().get(0).getNetVolume()).isEqualTo(BigDecimal.valueOf(33.11));
        assertThat(response.get(1).getValues().get(0).getPrice()).isEqualTo(BigDecimal.valueOf(33.22));
    }

}

File: kotklin_apx/src/test/kotlin/net/sympower/citizen/apx/DataResponseServiceTest.kt
package net.sympower.citizen.apx

import io.mockk.every
import io.mockk.mockk
import net.sympower.citizen.apx.config.ProviderEndpointsProperty
import net.sympower.citizen.apx.service.DataRequestService
import net.sympower.citizen.apx.service.DataResponseService
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import java.math.BigDecimal

class DataResponseServiceTest {

    private val mockDataRequestService = mockk<DataRequestService>()
    private val mockProviderEndpoints = mockk<ProviderEndpointsProperty>()
    private val dataResponseService = DataResponseService(mockDataRequestService,mockProviderEndpoints)


    @BeforeEach
    fun setUp() {
        every { mockProviderEndpoints.someProviderGet } returns "SOME_MOCK_URL"
    }

    @Test
    fun `should get DataResponse for the same quotes`() {
        // GIVEN
        every { mockDataRequestService.getProviderData(any()) } returns MockData.getSameQuotes()
        // WHEN
        val response = dataResponseService.getDataResponse()
        // THEN
        assertThat(response).hasSize(1)
        assertThat(response[0].market).isEqualTo("SomeMarket")
        assertThat(response[0].date).isEqualTo(1573599600000)
        assertThat(response[0].values).hasSize(2)
        assertThat(response[0].values[0].hour).isEqualTo(1)
        assertThat(response[0].values[0].netVolume).isEqualTo(BigDecimal.valueOf(11.11))
        assertThat(response[0].values[0].price).isEqualTo(BigDecimal.valueOf(11.22))
        assertThat(response[0].values[1].hour).isEqualTo(2)
        assertThat(response[0].values[1].netVolume).isEqualTo(BigDecimal.valueOf(22.11))
        assertThat(response[0].values[1].price).isEqualTo(BigDecimal.valueOf(22.22))
    }

    @Test
    fun `should get DataResponse for different quotes`() {
        // GIVEN
        every { mockDataRequestService.getProviderData(any()) } returns MockData.getDifferentQuotes()
        // WHEN
        val response = dataResponseService.getDataResponse()
        // THEN
        println("finish")
        assertThat(response).hasSize(2)
        assertThat(response[0].market).isEqualTo("SomeMarket")
        assertThat(response[0].date).isEqualTo(1573599600000)
        assertThat(response[0].values).hasSize(1)
        assertThat(response[0].values[0].hour).isEqualTo(1)
        assertThat(response[0].values[0].netVolume).isEqualTo(BigDecimal.valueOf(11.11))
        assertThat(response[0].values[0].price).isEqualTo(BigDecimal.valueOf(11.22))
        assertThat(response[1].market).isEqualTo("OtherMarket")
        assertThat(response[1].date).isEqualTo(1573599600000)
        assertThat(response[1].values).hasSize(1)
        assertThat(response[1].values[0].hour).isEqualTo(3)
        assertThat(response[1].values[0].netVolume).isEqualTo(BigDecimal.valueOf(33.11))
        assertThat(response[1].values[0].price).isEqualTo(BigDecimal.valueOf(33.22))
    }

}